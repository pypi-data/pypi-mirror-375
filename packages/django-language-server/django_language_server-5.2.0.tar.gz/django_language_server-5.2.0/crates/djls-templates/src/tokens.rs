use serde::Serialize;

#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum TokenType {
    Comment(String, String, Option<String>),
    DjangoBlock(String),
    DjangoVariable(String),
    Eof,
    HtmlTagOpen(String),
    HtmlTagClose(String),
    HtmlTagVoid(String),
    Newline,
    ScriptTagOpen(String),
    ScriptTagClose(String),
    StyleTagOpen(String),
    StyleTagClose(String),
    Text(String),
    Whitespace(usize),
}

impl TokenType {
    pub fn len(&self) -> usize {
        match self {
            TokenType::DjangoBlock(s)
            | TokenType::DjangoVariable(s)
            | TokenType::HtmlTagOpen(s)
            | TokenType::HtmlTagClose(s)
            | TokenType::HtmlTagVoid(s)
            | TokenType::ScriptTagOpen(s)
            | TokenType::ScriptTagClose(s)
            | TokenType::StyleTagOpen(s)
            | TokenType::StyleTagClose(s)
            | TokenType::Text(s) => s.len(),
            TokenType::Comment(content, _, _) => content.len(),
            TokenType::Whitespace(n) => *n,
            TokenType::Newline => 1,
            TokenType::Eof => 0,
        }
    }
}

#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Token {
    #[allow(clippy::struct_field_names)]
    token_type: TokenType,
    line: usize,
    start: Option<usize>,
}

impl Token {
    pub fn new(token_type: TokenType, line: usize, start: Option<usize>) -> Self {
        Self {
            token_type,
            line,
            start,
        }
    }

    pub fn lexeme(&self) -> String {
        match &self.token_type {
            TokenType::Comment(_, start, end) => match end {
                Some(end) => format!("{} {} {}", start, self.content(), end),
                None => format!("{} {}", start, self.content()),
            },
            TokenType::DjangoBlock(_) => format!("{{% {} %}}", self.content()),
            TokenType::DjangoVariable(_) => format!("{{{{ {} }}}}", self.content()),
            TokenType::Eof => String::new(),
            TokenType::HtmlTagOpen(_)
            | TokenType::ScriptTagOpen(_)
            | TokenType::StyleTagOpen(_) => format!("<{}>", self.content()),
            TokenType::HtmlTagClose(_)
            | TokenType::StyleTagClose(_)
            | TokenType::ScriptTagClose(_) => format!("</{}>", self.content()),
            TokenType::HtmlTagVoid(_) => format!("<{}/>", self.content()),
            TokenType::Newline | TokenType::Text(_) | TokenType::Whitespace(_) => self.content(),
        }
    }

    pub fn content(&self) -> String {
        match &self.token_type {
            TokenType::Comment(s, _, _)
            | TokenType::DjangoBlock(s)
            | TokenType::DjangoVariable(s)
            | TokenType::Text(s)
            | TokenType::HtmlTagOpen(s)
            | TokenType::HtmlTagClose(s)
            | TokenType::HtmlTagVoid(s)
            | TokenType::ScriptTagOpen(s)
            | TokenType::ScriptTagClose(s)
            | TokenType::StyleTagOpen(s)
            | TokenType::StyleTagClose(s) => s.to_string(),
            TokenType::Whitespace(len) => " ".repeat(*len),
            TokenType::Newline => "\n".to_string(),
            TokenType::Eof => String::new(),
        }
    }

    pub fn token_type(&self) -> &TokenType {
        &self.token_type
    }

    pub fn line(&self) -> &usize {
        &self.line
    }

    pub fn start(&self) -> Option<u32> {
        self.start
            .map(|s| u32::try_from(s).expect("Start position should fit in u32"))
    }

    pub fn length(&self) -> u32 {
        u32::try_from(self.token_type.len()).expect("Token length should fit in u32")
    }

    pub fn is_token_type(&self, token_type: &TokenType) -> bool {
        &self.token_type == token_type
    }
}

#[salsa::tracked]
pub struct TokenStream<'db> {
    #[tracked]
    #[returns(ref)]
    pub stream: Vec<Token>,
}

impl<'db> TokenStream<'db> {
    /// Check if the token stream is empty
    pub fn is_empty(self, db: &'db dyn crate::db::Db) -> bool {
        self.stream(db).is_empty()
    }

    /// Get the number of tokens
    pub fn len(self, db: &'db dyn crate::db::Db) -> usize {
        self.stream(db).len()
    }
}
