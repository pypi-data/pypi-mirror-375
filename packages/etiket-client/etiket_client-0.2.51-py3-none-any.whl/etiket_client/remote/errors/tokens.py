from etiket_client.remote.errors.meta import EtiketExceptionsMeta

class InvalidRequest(Exception):
    """The request is missing a required parameter, includes an invalid parameter value, or is otherwise malformed."""

class InvalidClient(Exception):
    """Client authentication failed."""

class InvalidGrant(Exception):
    """The provided authorization grant is invalid, expired, revoked, or was issued to another client."""

class UnauthorizedClient(Exception):
    """The authenticated client is not authorized to use this authorization grant type."""

class UnsupportedGrantType(Exception):
    """The authorization grant type is not supported by the authorization server."""

class InvalidScope(Exception):
    """The requested scope is invalid, unknown, malformed, or exceeds the scope granted."""

class AccessDenied(Exception):
    """The resource owner or authorization server denied the request."""

class UnsupportedResponseType(Exception):
    """The authorization server does not support obtaining an authorization code using this method."""

class TemporarilyUnavailable(Exception):
    """The authorization server is currently unable to handle the request."""

class TokenInvalid(Exception):
    """The provided token is invalid."""

class TokenRevoked(Exception):
    """The provided token has been revoked."""

class TokenUserCreateFailed(Exception):
    """Failed to create user from token information."""

class OAuth2Errors(metaclass=EtiketExceptionsMeta):
    invalid_request = InvalidRequest
    invalid_client = InvalidClient
    invalid_grant = InvalidGrant
    unauthorized_client = UnauthorizedClient
    unsupported_grant_type = UnsupportedGrantType
    invalid_scope = InvalidScope
    access_denied = AccessDenied
    unsupported_response_type = UnsupportedResponseType
    temporarily_unavailable = TemporarilyUnavailable

class TokenAuthErrors(metaclass=EtiketExceptionsMeta):
    token_invalid = TokenInvalid
    token_revoked = TokenRevoked
    token_user_create_failed = TokenUserCreateFailed