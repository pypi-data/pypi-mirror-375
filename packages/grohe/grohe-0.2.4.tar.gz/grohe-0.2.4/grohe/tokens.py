import httpx
import urllib.parse
from bs4 import BeautifulSoup

from grohe.dto.grohe_dto import GroheTokensDTO


class GroheTokens:
    def __init__(self, httpx_client: httpx.AsyncClient, api_url: str):
        self.__client = httpx_client
        self.__api_url = api_url


    async def get_tokens_from_credentials(self, grohe_email: str, grohe_password: str) -> GroheTokensDTO:
        """
        Get the initial access and refresh tokens from the given Grohe credentials.
        Args:
            grohe_email: The Grohe email.
            grohe_password: The Grohe password.

        Returns: A dict with the tokens.
        """
        response = await self.__client.get(f'{self.__api_url}/oidc/login', follow_redirects=True)
        response.raise_for_status()
        html = response.text

        soup = BeautifulSoup(html, 'html.parser')
        form = soup.find('form')
        if not form or 'action' not in form.attrs:
            raise Exception('Login form target URL not found')

        action_url = urllib.parse.urljoin(f'{self.__api_url}/oidc/login', form['action'])

        payload = {
            'username': grohe_email,
            'password': grohe_password,
        }

        headers = {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Referer': f'{self.__api_url}/oidc/login',
        }

        try:
            response = await self.__client.post(
                action_url, data=payload, headers=headers, follow_redirects=False
            )
            response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            if exc.response.status_code == 302:
                location = exc.response.headers.get('Location')
                if not location:
                    raise Exception('No redirect location found after login')

                tokens_url = location.replace('ondus://', 'https://')

                response = await self.__client.get(tokens_url)
                response.raise_for_status()
                json_data = response.json()

                tokens = GroheTokensDTO.from_dict(json_data)
                return tokens
            else:
                raise

        raise Exception(
            'Invalid username/password or unexpected response from Grohe service'
        )

    async def get_refresh_tokens(self, refresh_token: str) -> GroheTokensDTO:
        """
        Refresh the access and refresh tokens.
        Args:
            refresh_token: The refresh token.

        Returns: A dict with the new tokens.
        """
        data = {'refresh_token': refresh_token, 'grant_type': 'refresh_token'}
        response = await self.__client.post(f'{self.__api_url}/oidc/refresh', json=data)
        response.raise_for_status()
        json_data = response.json()

        tokens = GroheTokensDTO.from_dict(json_data)
        return tokens
