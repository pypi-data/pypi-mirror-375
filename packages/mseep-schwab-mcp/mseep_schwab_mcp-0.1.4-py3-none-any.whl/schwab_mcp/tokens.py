#

import json
import pathlib
import os
from typing import Any, Callable

import yaml
from platformdirs import user_data_dir


def token_path(app_name: str, filename: str = "token.yaml") -> str:
    """Get the path to the token file.

    This function returns the path to the token file based on the application name
    and the filename. The token file is stored in the user data directory.

    Args:
        app_name: The application name
        filename: The token file name

    Returns:
        The path to the token file
    """
    data_dir = user_data_dir(app_name)
    pathlib.Path(data_dir).mkdir(parents=True, exist_ok=True)
    return os.path.join(data_dir, filename)


def token_writer(token_path: str) -> Callable[[dict[str, Any], ...], None]:
    """Create a function that writes token data to a file.

    This function creates a token writer that supports both JSON and YAML formats
    based on the file extension. If the filename ends with '.json', JSON format
    will be used; otherwise, YAML format will be used.

    Args:
        token_path: Path to the token file

    Returns:
        A function that takes a token dictionary and writes it to the file
    """

    def write_token(token: dict[str, Any], *args, **kwargs) -> None:
        """Write the token data to a file.

        Args:
            token: The OAuth token data dictionary
            *args: Additional arguments (ignored)
            **kwargs: Additional keyword arguments (ignored)
        """
        if not token:
            return

        with open(token_path, "w") as f:
            if token_path.endswith(".json"):
                return json.dump(token, f)

            # Round Trip the token through JSON to ensure it's serializable
            return yaml.safe_dump(
                json.loads(json.dumps(token)),
                f,
                default_flow_style=False,
                explicit_start=True,
            )

    return write_token


def token_loader(token_path: str) -> Callable[[], dict[str, Any]]:
    """Create a function that loads token data from a file.

    This function creates a token loader that supports both JSON and YAML formats
    based on the file extension. If the filename ends with '.json', JSON format
    will be used; otherwise, YAML format will be used.

    Args:
        token_path: Path to the token file

    Returns:
        A function that loads and returns token data from the file
    """

    def load_token() -> dict[str, Any]:
        """Load the token data from a file.

        Returns:
            The OAuth token data as a dictionary
        """
        with open(token_path, "r") as f:
            if token_path.endswith(".json"):
                return json.load(f)

            return yaml.safe_load(f)

    return load_token


class Manager:
    def __init__(self, path: str):
        self.path = path
        self.load = token_loader(self.path)
        self.write = token_writer(self.path)

    def exists(self) -> bool:
        return os.path.exists(self.path)
