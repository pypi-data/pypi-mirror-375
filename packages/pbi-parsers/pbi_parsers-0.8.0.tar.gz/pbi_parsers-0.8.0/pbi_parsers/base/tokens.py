from dataclasses import dataclass, field
from typing import Any


@dataclass
class TextSlice:
    full_text: str = ""
    start: int = -1
    end: int = -1

    def __eq__(self, other: object) -> bool:
        """Checks equality based on the text slice."""
        if not isinstance(other, TextSlice):
            return NotImplemented
        return self.full_text == other.full_text and self.start == other.start and self.end == other.end

    def __hash__(self) -> int:
        """Returns a hash based on the text slice."""
        return hash((self.full_text, self.start, self.end))

    def __repr__(self) -> str:
        """Returns a string representation of the TextSlice."""
        return f"TextSlice(text='{self.get_text()}', start={self.start}, end={self.end})"

    def get_text(self) -> str:
        """Returns the text slice."""
        return self.full_text[self.start : self.end]


@dataclass
class BaseToken:
    tok_type: Any
    text_slice: TextSlice = field(default_factory=TextSlice)

    def __eq__(self, other: object) -> bool:
        """Checks equality based on token type and text slice."""
        if not isinstance(other, BaseToken):
            return NotImplemented
        return self.tok_type == other.tok_type and self.text_slice == other.text_slice

    def __hash__(self) -> int:
        """Returns a hash based on token type and text slice."""
        return hash((self.tok_type, self.text_slice))

    def __repr__(self) -> str:
        pretty_text = self.text_slice.get_text().replace("\n", "\\n").replace("\r", "\\r")
        return f"Token(type={self.tok_type.name}, text='{pretty_text}')"

    def position(self) -> tuple[int, int]:
        """Returns the start and end positions of the token.

        Returns:
            tuple[int, int]: A tuple containing the start and end positions of the token within the source text.

        """
        return self.text_slice.start, self.text_slice.end

    @property
    def text(self) -> str:
        """Returns the text underlying the token.

        Returns:
            str: The text of the token as a string.

        """
        return self.text_slice.get_text()
