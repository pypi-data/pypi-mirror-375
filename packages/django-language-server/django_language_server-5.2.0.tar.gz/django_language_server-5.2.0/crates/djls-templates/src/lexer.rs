use thiserror::Error;

use crate::tokens::Token;
use crate::tokens::TokenType;

pub struct Lexer {
    source: String,
    chars: Vec<char>,
    start: usize,
    current: usize,
    line: usize,
}

impl Lexer {
    #[must_use]
    pub fn new(source: &str) -> Self {
        Lexer {
            source: String::from(source),
            chars: source.chars().collect(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    #[allow(clippy::too_many_lines)]
    pub fn tokenize(&mut self) -> Result<Vec<Token>, LexerError> {
        let mut tokens = Vec::new();

        while !self.is_at_end() {
            self.start = self.current;

            let token_type = match self.peek()? {
                '{' => match self.peek_next()? {
                    '%' => {
                        self.consume_n(2)?; // {%
                        let content = self.consume_until("%}")?;
                        self.consume_n(2)?; // %}
                        TokenType::DjangoBlock(content)
                    }
                    '{' => {
                        self.consume_n(2)?; // {{
                        let content = self.consume_until("}}")?;
                        self.consume_n(2)?; // }}
                        TokenType::DjangoVariable(content)
                    }
                    '#' => {
                        self.consume_n(2)?; // {#
                        let content = self.consume_until("#}")?;
                        self.consume_n(2)?; // #}
                        TokenType::Comment(content, "{#".to_string(), Some("#}".to_string()))
                    }
                    _ => {
                        self.consume()?; // {
                        TokenType::Text(String::from("{"))
                    }
                },

                '<' => match self.peek_next()? {
                    '/' => {
                        self.consume_n(2)?; // </
                        let tag = self.consume_until(">")?;
                        self.consume()?; // >
                        TokenType::HtmlTagClose(tag)
                    }
                    '!' if self.matches("<!--") => {
                        self.consume_n(4)?; // <!--
                        let content = self.consume_until("-->")?;
                        self.consume_n(3)?; // -->
                        TokenType::Comment(content, "<!--".to_string(), Some("-->".to_string()))
                    }
                    _ => {
                        self.consume()?; // consume <
                        let tag = self.consume_until(">")?;
                        self.consume()?; // consume >
                        if tag.starts_with("script") {
                            TokenType::ScriptTagOpen(tag)
                        } else if tag.starts_with("style") {
                            TokenType::StyleTagOpen(tag)
                        } else if tag.ends_with('/') {
                            TokenType::HtmlTagVoid(tag.trim_end_matches('/').to_string())
                        } else {
                            TokenType::HtmlTagOpen(tag)
                        }
                    }
                },

                '/' => match self.peek_next()? {
                    '/' => {
                        self.consume_n(2)?; // //
                        let content = self.consume_until("\n")?;
                        TokenType::Comment(content, "//".to_string(), None)
                    }
                    '*' => {
                        self.consume_n(2)?; // /*
                        let content = self.consume_until("*/")?;
                        self.consume_n(2)?; // */
                        TokenType::Comment(content, "/*".to_string(), Some("*/".to_string()))
                    }
                    _ => {
                        self.consume()?;
                        TokenType::Text("/".to_string())
                    }
                },

                c if c.is_whitespace() => {
                    if c == '\n' || c == '\r' {
                        self.consume()?; // \r or \n
                        if c == '\r' && self.peek()? == '\n' {
                            self.consume()?; // \n of \r\n
                        }
                        TokenType::Newline
                    } else {
                        self.consume()?; // Consume the first whitespace
                        while !self.is_at_end() && self.peek()?.is_whitespace() {
                            if self.peek()? == '\n' || self.peek()? == '\r' {
                                break;
                            }
                            self.consume()?;
                        }
                        let whitespace_count = self.current - self.start;
                        TokenType::Whitespace(whitespace_count)
                    }
                }

                _ => {
                    let mut text = String::new();
                    while !self.is_at_end() {
                        let c = self.peek()?;
                        if c == '{' || c == '<' || c == '\n' {
                            break;
                        }
                        text.push(c);
                        self.consume()?;
                    }
                    TokenType::Text(text)
                }
            };

            let token = Token::new(token_type, self.line, Some(self.start));

            match self.peek_previous()? {
                '\n' => self.line += 1,
                '\r' => {
                    self.line += 1;
                    if self.peek()? == '\n' {
                        self.current += 1;
                    }
                }
                _ => {}
            }

            tokens.push(token);
        }

        // Add EOF token
        let eof_token = Token::new(TokenType::Eof, self.line, None);
        tokens.push(eof_token);

        Ok(tokens)
    }

    fn peek(&self) -> Result<char, LexerError> {
        self.peek_at(0)
    }

    fn peek_next(&self) -> Result<char, LexerError> {
        self.peek_at(1)
    }

    fn peek_previous(&self) -> Result<char, LexerError> {
        self.peek_at(-1)
    }

    #[allow(dead_code)]
    fn peek_until(&self, end: &str) -> bool {
        let mut index = self.current;
        let end_chars: Vec<char> = end.chars().collect();

        while index < self.chars.len() {
            if self.chars[index..].starts_with(&end_chars) {
                return true;
            }
            index += 1;
        }
        false
    }

    #[allow(clippy::cast_sign_loss)]
    fn peek_at(&self, offset: isize) -> Result<char, LexerError> {
        // Safely handle negative offsets
        let index = if offset < 0 {
            // Check if we would underflow
            if self.current < offset.unsigned_abs() {
                return Err(LexerError::AtBeginningOfSource);
            }
            self.current - offset.unsigned_abs()
        } else {
            // Safe addition since offset is positive
            self.current + (offset as usize)
        };

        self.item_at(index)
    }

    fn item_at(&self, index: usize) -> Result<char, LexerError> {
        if index >= self.source.len() {
            // Return a null character when past the end, a bit of a departure from
            // idiomatic Rust code, but makes writing the matching above and testing
            // much easier
            Ok('\0')
        } else {
            self.source
                .chars()
                .nth(index)
                .ok_or(LexerError::InvalidCharacterAccess)
        }
    }

    fn matches(&mut self, pattern: &str) -> bool {
        let mut i = self.current;
        for c in pattern.chars() {
            if i >= self.chars.len() || self.chars[i] != c {
                return false;
            }
            i += 1;
        }
        true
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn consume(&mut self) -> Result<char, LexerError> {
        if self.is_at_end() {
            return Err(LexerError::AtEndOfSource);
        }
        self.current += 1;
        self.peek_previous()
    }

    fn consume_n(&mut self, count: usize) -> Result<String, LexerError> {
        let start = self.current;
        for _ in 0..count {
            self.consume()?;
        }
        Ok(self.source[start..self.current].trim().to_string())
    }

    #[allow(dead_code)]
    fn consume_chars(&mut self, s: &str) -> Result<char, LexerError> {
        for c in s.chars() {
            if c != self.peek()? {
                return Err(LexerError::UnexpectedCharacter(c, self.line));
            }
            self.consume()?;
        }
        self.peek_previous()
    }

    fn consume_until(&mut self, s: &str) -> Result<String, LexerError> {
        let start = self.current;
        while !self.is_at_end() {
            if self.chars[self.current..self.chars.len()]
                .starts_with(s.chars().collect::<Vec<_>>().as_slice())
            {
                return Ok(self.source[start..self.current].trim().to_string());
            }
            self.consume()?;
        }
        Err(LexerError::UnexpectedEndOfInput)
    }
}

#[derive(Error, Debug)]
pub enum LexerError {
    #[error("empty token at line {0}")]
    EmptyToken(usize),

    #[error("unexpected character '{0}' at line {1}")]
    UnexpectedCharacter(char, usize),

    #[error("unexpected end of input")]
    UnexpectedEndOfInput,

    #[error("source is empty")]
    EmptySource,

    #[error("at beginning of source")]
    AtBeginningOfSource,

    #[error("at end of source")]
    AtEndOfSource,

    #[error("invalid character access")]
    InvalidCharacterAccess,

    #[error("unexpected token type '{0:?}'")]
    UnexpectedTokenType(TokenType),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tokenize_html() {
        let source = r#"<div class="container" id="main" disabled></div>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_django_variable() {
        let source = "{{ user.name|default:\"Anonymous\"|title }}";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_django_block() {
        let source = "{% if user.is_staff %}Admin{% else %}User{% endif %}";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_comments() {
        let source = r"<!-- HTML comment -->
{# Django comment #}
<script>
    // JS single line comment
    /* JS multi-line
       comment */
</script>
<style>
    /* CSS comment */
</style>";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_script() {
        let source = r#"<script type="text/javascript">
    // Single line comment
    const x = 1;
    /* Multi-line
       comment */
    console.log(x);
</script>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_style() {
        let source = r#"<style type="text/css">
    /* Header styles */
    .header {
        color: blue;
    }
</style>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_error_cases() {
        // Unterminated tokens
        assert!(Lexer::new("{{ user.name").tokenize().is_err()); // No closing }}
        assert!(Lexer::new("{% if").tokenize().is_err()); // No closing %}
        assert!(Lexer::new("{#").tokenize().is_err()); // No closing #}
        assert!(Lexer::new("<div").tokenize().is_err()); // No closing >

        // Invalid characters or syntax within tokens
        assert!(Lexer::new("{{}}").tokenize().is_ok()); // Empty but valid
        assert!(Lexer::new("{%  %}").tokenize().is_ok()); // Empty but valid
        assert!(Lexer::new("{##}").tokenize().is_ok()); // Empty but valid
    }

    #[test]
    fn test_tokenize_nested_delimiters() {
        let source = r"{{ user.name }}
{% if true %}
{# comment #}
<!-- html comment -->
<div>text</div>";
        assert!(Lexer::new(source).tokenize().is_ok());
    }

    #[test]
    fn test_tokenize_everything() {
        let source = r#"<!DOCTYPE html>
<html>
<head>
    <style type="text/css">
        /* Style header */
        .header { color: blue; }
    </style>
    <script type="text/javascript">
        // Init app
        const app = {
            /* Config */
            debug: true
        };
    </script>
</head>
<body>
    <!-- Header section -->
    <div class="header" id="main" data-value="123" disabled>
        {% if user.is_authenticated %}
            {# Welcome message #}
            <h1>Welcome, {{ user.name|default:"Guest"|title }}!</h1>
            {% if user.is_staff %}
                <span>Admin</span>
            {% else %}
                <span>User</span>
            {% endif %}
        {% endif %}
    </div>
</body>
</html>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }
}
