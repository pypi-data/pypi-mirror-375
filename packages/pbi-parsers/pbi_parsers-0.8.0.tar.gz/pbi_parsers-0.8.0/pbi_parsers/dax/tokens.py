from dataclasses import dataclass
from enum import Enum, auto

from pbi_parsers.base import BaseToken


class TokenType(Enum):
    AMPERSAND_OPERATOR = auto()
    ASC = auto()
    BRACKETED_IDENTIFIER = auto()
    COMMA = auto()
    COMPARISON_OPERATOR = auto()
    DESC = auto()
    DIVIDE_SIGN = auto()
    DOUBLE_AMPERSAND_OPERATOR = auto()
    DOUBLE_PIPE_OPERATOR = auto()
    EOF = auto()
    EQUAL_SIGN = auto()
    EXPONENTIATION_SIGN = auto()
    FALSE = auto()
    IN = auto()
    LEFT_CURLY_BRACE = auto()
    LEFT_PAREN = auto()
    MINUS_SIGN = auto()
    MODULUS_SIGN = auto()
    MULTIPLY_SIGN = auto()
    MULTI_LINE_COMMENT = auto()
    NOT_EQUAL_SIGN = auto()
    NUMBER_LITERAL = auto()
    PERIOD = auto()
    PLUS_SIGN = auto()
    RETURN = auto()
    RIGHT_CURLY_BRACE = auto()
    RIGHT_PAREN = auto()
    SINGLE_LINE_COMMENT = auto()
    SINGLE_QUOTED_IDENTIFIER = auto()
    STRING_LITERAL = auto()
    TRUE = auto()
    UNQUOTED_IDENTIFIER = auto()
    VARIABLE = auto()
    WHITESPACE = auto()


KEYWORD_MAPPING = {
    "TRUE": TokenType.TRUE,
    "FALSE": TokenType.FALSE,
    "ASC": TokenType.ASC,
    "DESC": TokenType.DESC,
}


@dataclass
class Token(BaseToken):
    tok_type: TokenType = TokenType.EOF
