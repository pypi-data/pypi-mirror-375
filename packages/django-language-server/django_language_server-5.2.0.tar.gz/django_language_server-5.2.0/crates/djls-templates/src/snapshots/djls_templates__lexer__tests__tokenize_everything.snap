---
source: crates/djls-templates/src/lexer.rs
expression: tokens
---
- token_type:
    HtmlTagOpen: "!DOCTYPE html"
  line: 1
  start: 0
- token_type: Newline
  line: 1
  start: 15
- token_type:
    HtmlTagOpen: html
  line: 2
  start: 16
- token_type: Newline
  line: 2
  start: 22
- token_type:
    HtmlTagOpen: head
  line: 3
  start: 23
- token_type: Newline
  line: 3
  start: 29
- token_type:
    Whitespace: 4
  line: 4
  start: 30
- token_type:
    StyleTagOpen: "style type=\"text/css\""
  line: 4
  start: 34
- token_type: Newline
  line: 4
  start: 57
- token_type:
    Whitespace: 8
  line: 5
  start: 58
- token_type:
    Comment:
      - Style header
      - /*
      - "*/"
  line: 5
  start: 66
- token_type: Newline
  line: 5
  start: 84
- token_type:
    Whitespace: 8
  line: 6
  start: 85
- token_type:
    Text: ".header "
  line: 6
  start: 93
- token_type:
    Text: "{"
  line: 6
  start: 101
- token_type:
    Whitespace: 1
  line: 6
  start: 102
- token_type:
    Text: "color: blue; }"
  line: 6
  start: 103
- token_type: Newline
  line: 6
  start: 117
- token_type:
    Whitespace: 4
  line: 7
  start: 118
- token_type:
    HtmlTagClose: style
  line: 7
  start: 122
- token_type: Newline
  line: 7
  start: 130
- token_type:
    Whitespace: 4
  line: 8
  start: 131
- token_type:
    ScriptTagOpen: "script type=\"text/javascript\""
  line: 8
  start: 135
- token_type: Newline
  line: 8
  start: 166
- token_type:
    Whitespace: 8
  line: 9
  start: 167
- token_type:
    Comment:
      - Init app
      - //
      - ~
  line: 9
  start: 175
- token_type: Newline
  line: 9
  start: 186
- token_type:
    Whitespace: 8
  line: 10
  start: 187
- token_type:
    Text: "const app = "
  line: 10
  start: 195
- token_type:
    Text: "{"
  line: 10
  start: 207
- token_type: Newline
  line: 10
  start: 208
- token_type:
    Whitespace: 12
  line: 11
  start: 209
- token_type:
    Comment:
      - Config
      - /*
      - "*/"
  line: 11
  start: 221
- token_type: Newline
  line: 11
  start: 233
- token_type:
    Whitespace: 12
  line: 12
  start: 234
- token_type:
    Text: "debug: true"
  line: 12
  start: 246
- token_type: Newline
  line: 12
  start: 257
- token_type:
    Whitespace: 8
  line: 13
  start: 258
- token_type:
    Text: "};"
  line: 13
  start: 266
- token_type: Newline
  line: 13
  start: 268
- token_type:
    Whitespace: 4
  line: 14
  start: 269
- token_type:
    HtmlTagClose: script
  line: 14
  start: 273
- token_type: Newline
  line: 14
  start: 282
- token_type:
    HtmlTagClose: head
  line: 15
  start: 283
- token_type: Newline
  line: 15
  start: 290
- token_type:
    HtmlTagOpen: body
  line: 16
  start: 291
- token_type: Newline
  line: 16
  start: 297
- token_type:
    Whitespace: 4
  line: 17
  start: 298
- token_type:
    Comment:
      - Header section
      - "<!--"
      - "-->"
  line: 17
  start: 302
- token_type: Newline
  line: 17
  start: 325
- token_type:
    Whitespace: 4
  line: 18
  start: 326
- token_type:
    HtmlTagOpen: "div class=\"header\" id=\"main\" data-value=\"123\" disabled"
  line: 18
  start: 330
- token_type: Newline
  line: 18
  start: 386
- token_type:
    Whitespace: 8
  line: 19
  start: 387
- token_type:
    DjangoBlock: if user.is_authenticated
  line: 19
  start: 395
- token_type: Newline
  line: 19
  start: 425
- token_type:
    Whitespace: 12
  line: 20
  start: 426
- token_type:
    Comment:
      - Welcome message
      - "{#"
      - "#}"
  line: 20
  start: 438
- token_type: Newline
  line: 20
  start: 459
- token_type:
    Whitespace: 12
  line: 21
  start: 460
- token_type:
    HtmlTagOpen: h1
  line: 21
  start: 472
- token_type:
    Text: "Welcome, "
  line: 21
  start: 476
- token_type:
    DjangoVariable: "user.name|default:\"Guest\"|title"
  line: 21
  start: 485
- token_type:
    Text: "!"
  line: 21
  start: 522
- token_type:
    HtmlTagClose: h1
  line: 21
  start: 523
- token_type: Newline
  line: 21
  start: 528
- token_type:
    Whitespace: 12
  line: 22
  start: 529
- token_type:
    DjangoBlock: if user.is_staff
  line: 22
  start: 541
- token_type: Newline
  line: 22
  start: 563
- token_type:
    Whitespace: 16
  line: 23
  start: 564
- token_type:
    HtmlTagOpen: span
  line: 23
  start: 580
- token_type:
    Text: Admin
  line: 23
  start: 586
- token_type:
    HtmlTagClose: span
  line: 23
  start: 591
- token_type: Newline
  line: 23
  start: 598
- token_type:
    Whitespace: 12
  line: 24
  start: 599
- token_type:
    DjangoBlock: else
  line: 24
  start: 611
- token_type: Newline
  line: 24
  start: 621
- token_type:
    Whitespace: 16
  line: 25
  start: 622
- token_type:
    HtmlTagOpen: span
  line: 25
  start: 638
- token_type:
    Text: User
  line: 25
  start: 644
- token_type:
    HtmlTagClose: span
  line: 25
  start: 648
- token_type: Newline
  line: 25
  start: 655
- token_type:
    Whitespace: 12
  line: 26
  start: 656
- token_type:
    DjangoBlock: endif
  line: 26
  start: 668
- token_type: Newline
  line: 26
  start: 679
- token_type:
    Whitespace: 8
  line: 27
  start: 680
- token_type:
    DjangoBlock: endif
  line: 27
  start: 688
- token_type: Newline
  line: 27
  start: 699
- token_type:
    Whitespace: 4
  line: 28
  start: 700
- token_type:
    HtmlTagClose: div
  line: 28
  start: 704
- token_type: Newline
  line: 28
  start: 710
- token_type:
    HtmlTagClose: body
  line: 29
  start: 711
- token_type: Newline
  line: 29
  start: 718
- token_type:
    HtmlTagClose: html
  line: 30
  start: 719
- token_type: Eof
  line: 30
  start: ~
